plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go

```diff
@@ -250,89 +285,204 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
+	default:
+		return ""
+	}
+}
+
+// Recursively parse tag fields
+func (c *CiscoTelemetryMDT) parseKeyField(tags map[string]string, field *telemetry.TelemetryField, prefix string) {
+	localname := strings.Replace(field.Name, "-", "_", -1)
+	name := localname
+	if len(name) == 0 {
+		name = prefix
+	} else if len(prefix) > 0 {
+		name = prefix + "/" + localname
 	}
 
-	if value != nil {
-		// Distinguish between tags (keys) and fields (data) to write to
-		if fields != nil {
-			fields[namebuf.String()] = value
+	if tag := decodeTag(field); len(tag) > 0 {
+		localname := strings.Replace(field.Name, "-", "_", -1)
+		if _, exists := tags[localname]; !exists { // Use short keys whenever possible
+			tags[localname] = tag
 		} else {
-			if _, exists := tags[field.Name]; !exists { // Use short keys whenever possible
-				tags[field.Name] = fmt.Sprint(value)
-			} else {
-				tags[namebuf.String()] = fmt.Sprint(value)
+			tags[name] = tag
+		}
+	}
+
+	for _, subfield := range field.Fields {
+		c.parseKeyField(tags, subfield, name)
+	}
+}
+
+func (c *CiscoTelemetryMDT) parseContentField(grouper *metric.SeriesGrouper, field *telemetry.TelemetryField, prefix string,
+	path string, tags map[string]string, timestamp time.Time) {
+	name := strings.Replace(field.Name, "-", "_", -1)
+	if len(name) == 0 {
+		name = prefix
+	} else if len(prefix) > 0 {
+		name = prefix + "/" + name
+	}
+
+	extraTags := c.extraTags[path+"/"+name]
+
+	if value := decodeValue(field); value != nil {
+		// Do alias lookup, to shorten measurement names
+		measurement := path
+		if alias, ok := c.aliases[path]; ok {
+			measurement = alias
+		} else {
+			c.mutex.Lock()
+			if _, haveWarned := c.warned[path]; !haveWarned {
+				log.Printf("D! [inputs.cisco_telemetry_mdt]: No measurement alias for encoding path: %s", path)
+				c.warned[path] = struct{}{}
+			}
+			c.mutex.Unlock()
+		}
+
+		grouper.Add(measurement, tags, timestamp, name, value)
+	}
+
+	if len(extraTags) > 0 {
+		for _, subfield := range field.Fields {
+			if _, isExtraTag := extraTags[subfield.Name]; isExtraTag {
+				tags[name+"/"+subfield.Name] = decodeTag(subfield)
 			}
 		}
 	}
 
+	var nxAttributes, nxChildren, nxRows *telemetry.TelemetryField
+	isNXOS := !strings.ContainsRune(path, ':') // IOS-XR and IOS-XE have a colon in their encoding path, NX-OS does not
 	for _, subfield := range field.Fields {
-		c.parseGPBKVField(subfield, namebuf, path, timestamp, tags, fields)
+		if isNXOS && subfield.Name == "attributes" && len(subfield.Fields) > 0 {
+			nxAttributes = subfield.Fields[0]
+		} else if isNXOS && subfield.Name == "children" && len(subfield.Fields) > 0 {
+			nxChildren = subfield
+		} else if isNXOS && strings.HasPrefix(subfield.Name, "ROW_") {
+			nxRows = subfield
+		} else if _, isExtraTag := extraTags[subfield.Name]; !isExtraTag { // Regular telemetry decoding
+			c.parseContentField(grouper, subfield, name, path, tags, timestamp)
+		}
 	}
 
-	namebuf.Truncate(namelen)
+	if nxAttributes != nil {
```

Comment:
Return early if unset.

Reply:
Okay reorganized this part a bit, to unindent some longer code branches.